hudi-common/src/main/java/org/apache/hudi/common/table/read/HoodieBaseFileGroupRecordBuffer.java

```diff
@@ -278,14 +289,19 @@ protected Option<DeleteRecord> doProcessNextDeletedRecord(DeleteRecord deleteRec
           if (isDeleteRecordWithNaturalOrder(existingRecordMetadataPair.getLeft(), existingOrderingVal)) {
             return Option.empty();
           }
-          Comparable deleteOrderingVal = deleteRecord.getOrderingValue();
-          // Checks the ordering value does not equal to 0
-          // because we use 0 as the default value which means natural order
-          boolean chooseExisting = !deleteOrderingVal.equals(0)
-              && ReflectionUtils.isSameClass(existingOrderingVal, deleteOrderingVal)
-              && existingOrderingVal.compareTo(deleteOrderingVal) > 0;
+          Comparable deleteOrderingVal = readerContext.getOrderingValue(
+              Option.empty(), Collections.emptyMap(), readerSchema, orderingFieldName, orderingFieldType, orderingFieldDefault);
+          deleteOrderingVal = deleteRecord.getOrderingValue() == null ? deleteOrderingVal : deleteRecord.getOrderingValue();
```

Comment:
for now, if we are going to take a stand that delete will just override any previous value, why can't we just set the default ordering value (readerContext.castValue(0, orderingFieldType)) as the ordering value for delete records. and it make the code simple and easy to read.

Reply:
The default ordering value like (0) should be lower than the new inserted records with the same key, then the delete records are lost. Therefore, we have to use some extra flag to remember if a delete without valid ordering value exists.